fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,783 @@
+
+    compact_space.finish()
+}
+
+#[test]
+fn compact_space_test() {
+    let ips = vec![
+        2u128, 4u128, 1000, 1001, 1002, 1003, 1004, 1005, 1008, 1010, 1012, 1260,
+    ];
+    let compact_space = get_compact_space(&ips, 11);
+    assert_eq!(compact_space.null_value, 5);
+    let amplitude = compact_space.amplitude_compact_space();
+    assert_eq!(amplitude, 20);
+    assert_eq!(2, compact_space.to_compact(2).unwrap());
+    assert_eq!(compact_space.to_compact(100).unwrap_err(), 0);
+
+    let mut output = vec![];
+    compact_space.serialize(&mut output).unwrap();
+
+    assert_eq!(
+        compact_space,
+        CompactSpace::deserialize(&mut &output[..]).unwrap()
+    );
+
+    for ip in &ips {
+        let compact = compact_space.to_compact(*ip).unwrap();
+        assert_eq!(compact_space.unpack(compact), *ip);
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct CompactSpaceBuilder {
+    covered_space: Vec<RangeInclusive<u128>>,
+}
+
+impl CompactSpaceBuilder {
+    /// Creates a new compact space builder which will initially cover the whole space.
+    fn new() -> Self {
+        Self {
+            covered_space: vec![0..=u128::MAX],
+        }
+    }
+
+    /// Will extend the first range and assign the null value to it.
+    fn assign_and_return_null(&mut self) -> u128 {
+        self.covered_space[0] = *self.covered_space[0].start()..=*self.covered_space[0].end() + 1;
+        *self.covered_space[0].end()
+    }
+
+    /// Assumes that repeated add_blank calls don't overlap, which will be the case on sorted
+    /// values.
+    fn add_blank(&mut self, blank: RangeInclusive<u128>) {
+        let position = self
+            .covered_space
+            .iter()
+            .position(|range| range.start() <= blank.start() && range.end() >= blank.end());
+        if let Some(position) = position {
+            let old_range = self.covered_space.remove(position);
+            // Exact match, just remove
+            if old_range == blank {
+                return;
+            }
+            let new_range_end = blank.end().saturating_add(1)..=*old_range.end();
+            if old_range.start() == blank.start() {
+                self.covered_space.insert(position, new_range_end);
+                return;
+            }
+            let new_range_start = *old_range.start()..=blank.start().saturating_sub(1);
+            if old_range.end() == blank.end() {
+                self.covered_space.insert(position, new_range_start);
+                return;
+            }
+            self.covered_space.insert(position, new_range_end);
+            self.covered_space.insert(position, new_range_start);
+        }
+    }
+    fn finish(mut self) -> CompactSpace {
+        let null_value = self.assign_and_return_null();
+
+        let mut compact_start: u64 = 0;
+        let mut ranges_and_compact_start = vec![];
+        for cov in self.covered_space {
+            let covered_range_len = cov.end() - cov.start();
```

Comment:
why not `cov.end() - cov.start() + 1` ?
Range have a len via ExactIterator.

Reply:
ExactIterator is not implemented for u128